client/src/main/java/io/pravega/client/stream/ReaderGroup.java

```diff
@@ -137,13 +138,52 @@
+     * Returns a {@link StreamCut} for each stream that this reader group is reading from.
+     * The stream cut corresponds to the last updated read offsets of the readers, and
+     * it can be used by the application as reference to such a position.
+     * A more precise {@link StreamCut}, with the latest read offsets can be obtained using
+     * {@link ReaderGroup#generateStreamCuts(ScheduledExecutorService)} API.
      *
      * @return Map of streams that this group is reading from to the corresponding cuts.
      */
     Map<Stream, StreamCut> getStreamCuts();
+
+    /**
+     * Generates a {@link StreamCut} after co-ordinating with all the readers using
+     * {@link io.pravega.client.state.StateSynchronizer}. A {@link StreamCut} is
+     * generated by using the latest segment read offsets returned by the readers
+     * along with unassigned segments (if any).
+     *
+     * The configuration {@link ReaderGroupConfig#getGroupRefreshTimeMillis()} decides
+     * the maximum delay by which the readers return the latest read offsets of their
+     * assigned segments.
+     * <p>
+     * The {@link StreamCut} generated by this API can be used by the application as a
+     * reference to a position in the stream. This is guaranteed to be greater than or
+     * equal to the position of the readers at the point of invocation of the API. The
+     * {@link StreamCut}s generated can be used to perform bounded processing of the Stream
+     * by configuring a {@link ReaderGroup} with a {@link ReaderGroupConfig} where the
+     * {@link StreamCut}s are specified as the lower bound and/or upper bounds using the
+     * apis
+     * {@link ReaderGroupConfig.ReaderGroupConfigBuilder#stream(Stream, StreamCut, StreamCut)}
+     * or {@link ReaderGroupConfig.ReaderGroupConfigBuilder#stream(Stream, StreamCut)} or
+     * {@link ReaderGroupConfig.ReaderGroupConfigBuilder#startFromStreamCuts(Map)}.
+     * <p>
+     * Note: Generating a precise {@link StreamCut}, for example a {@link StreamCut} pointing to
+     * end of Q1 across all segments, is difficult as it depends on the configuration
+     * {@link ReaderGroupConfig#getGroupRefreshTimeMillis()} which decides the duration by
+     * which all the readers running on different machines/ processes respond with their
+     * latest read offsets. Hence, the {@link StreamCut} would point to a position in the
+     * {@link Stream} which might include events from Q2. The application thus would need to
+     * filter out such additional events.
+     *
+     * @param backgroundExecutor A thread pool that will be used to poll if the
+     *                           positions from all the readers have been fetched.
+     * @return A future to a Map of Streams (that this group is reading from) to
+     * its corresponding cuts.
+     */
+    @Beta
+    CompletableFuture<Map<Stream, StreamCut>> generateStreamCuts(ScheduledExecutorService backgroundExecutor);
```

Comment:
From an API perspective, it will be nice if it returns both `startStreamCut` and `endStreamcut` at the time of the execution where the `startStreamcut` being the current reader position and the `endStreamCut` being the current stream end position. This will provide the client applications with a bounded data set that it can operate on. For example, an application could execute `generateStreamCuts(...)` and store the start and end streamcut information and another application could make use of these bounded `streamcuts` to iterate on the data by using  `BatchClient.getSegments(stream, startStreamCut, endStreamCut)` API.

Reply:
The API `ReaderGroup#generateStreamCuts()` is used to generated a `StreamCut` based on the current position(s) of the reader(s) inside a given reader group. The  `StreamCut` pointing to the current `TAIL` of the stream can be fetched using `StreamManager#getStreamInfo`. 

Also, since the `StreamCut` pointing to the `TAIL` of a stream is every changing it is better if we  have a separate APIs to fetch the current and TAIL stream cuts.  ( note: `StreamManager#getStreamInfo` was promoted from the BatchClient to ensure it is easier for both the streaming and batch client users to fetch details of the stream).